internal/ihop/client.go

```diff
@@ -0,0 +1,488 @@
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/session/auth/authprovider"
+	"github.com/moby/buildkit/session/secrets/secretsprovider"
+	"github.com/moby/buildkit/util/progress/progressui"
+)
+
+type Image struct {
+	Tag          string
+	Digest       string
+	OS           string
+	Architecture string
+	User         string
+
+	Env    []string
+	Labels map[string]string
+
+	Layers []Layer
+}
+
+func (i Image) ToDaemonImage() (v1.Image, error) {
+	ref, err := name.ParseReference(i.Tag)
+	if err != nil {
+		return nil, err
+	}
+
+	image, err := daemon.Image(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	return image, nil
+}
+
+type Layer struct {
+	DiffID string
+
+	v1.Layer
+}
+
+type Client struct {
+	dir      string
+	docker   *docker.Client
+	keychain authn.Keychain
+}
+
+func NewClient(dir string) (Client, error) {
+	client, err := docker.NewClientWithOpts(docker.FromEnv)
+	if err != nil {
+		return Client{}, err
+	}
+
+	return Client{
+		dir:      dir,
+		docker:   client,
+		keychain: authn.DefaultKeychain,
+	}, nil
+}
+
+func (c Client) Build(def DefinitionImage, platform string) (Image, error) {
+	sum := sha256.Sum256([]byte(def.Dockerfile))
+	sess, err := session.NewSession(context.Background(), def.Dockerfile, hex.EncodeToString(sum[:]))
+	if err != nil {
+		return Image{}, err
+	}
+
+	dockerAuthProvider := authprovider.NewDockerAuthProvider(os.Stderr)
+	sess.Allow(dockerAuthProvider)
+
+	if len(def.Secrets) > 0 {
+		secretsDir, err := ioutil.TempDir("", "docker-secrets")
+		if err != nil {
+			return Image{}, err
+		}
+		defer os.RemoveAll(secretsDir)
+
+		fs := make([]secretsprovider.Source, 0, len(def.Secrets))
+		for id, secret := range def.Secrets {
+			path := filepath.Join(secretsDir, id)
+			err = os.WriteFile(path, []byte(secret), 0600)
+			if err != nil {
+				return Image{}, err
+			}
+
+			fs = append(fs, secretsprovider.Source{
+				ID:       id,
+				FilePath: path,
+			})
+		}
+
+		store, err := secretsprovider.NewStore(fs)
+		if err != nil {
+			return Image{}, err
+		}
+
+		sess.Allow(secretsprovider.NewSecretProvider(store))
+	}
+
+	go func() {
+		_ = sess.Run(context.Background(), func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+			return c.docker.DialHijack(ctx, "/session", proto, meta)
+		})
+	}()
+	defer sess.Close()
+
+	contextDir := filepath.Dir(def.Dockerfile)
+	excludes, err := build.ReadDockerignore(contextDir)
+	if err != nil {
+		return Image{}, err
+	}
+
+	contextDir, relDockerfile, err := build.GetContextFromLocalDir(contextDir, def.Dockerfile)
+	if err != nil {
+		return Image{}, err
+	}
+
+	buildContext, err := archive.TarWithOptions(contextDir, &archive.TarOptions{
+		ExcludePatterns: excludes,
+		ChownOpts:       &idtools.Identity{UID: 0, GID: 0},
+	})
+	if err != nil {
+		return Image{}, err
+	}
+	defer buildContext.Close()
+
+	tag, err := randomName()
+	if err != nil {
+		return Image{}, err
+	}
+
+	tag = fmt.Sprintf("paketo.io/stack/%s", tag)
```

Comment:
We have to come up with a reference for the image we are building. We'll want to avoid possibly conflicting with any images that are currently on the daemon and so we choose a randomly generated tag.